icicle-compiler/src/Icicle/Runtime/Data/Array.hs

```diff
@@ -297,9 +297,24 @@ grow pool arrayOld@(Array ptrOld) lengthNew = do
     !capacityNew =
       calculateCapacity lengthNew
 
-  if capacityOld >= capacityNew then do
+  if lengthOld > lengthNew then do
+    let
+      deadSize =
+        (lengthOld - lengthNew) * elementSize
+
+      deadPtr =
+        unsafeElementPtr arrayOld `plusPtr`
+        fromIntegral (lengthNew * elementSize)
+
+    c_memset deadPtr 0 (fromIntegral deadSize)
```

Comment:
Yeah, so I decided the easiest way to do it was to make sure the unused memory is always initialised to zero. So when we shrink from an array length of say `10` to a length of `5` we need to zero out indices `5` to `9`.